tests/cleanup_and_retention_test.go

```diff
@@ -1,191 +1,166 @@
 package icingadb_test
 
 import (
-	"context"
+	"encoding/binary"
 	"fmt"
-	"github.com/icinga/icingadb/pkg/icingadb"
-	"github.com/icinga/icingadb/pkg/icingadb/history"
-	"github.com/icinga/icingadb/pkg/logging"
+	"github.com/goccy/go-yaml"
+	"github.com/icinga/icinga-testing/services"
+	"github.com/icinga/icinga-testing/utils/eventually"
 	"github.com/jmoiron/sqlx"
-	"github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"sync"
 	"testing"
 	"time"
 )
 
 const MaxPlaceholders = 1 << 13
 
 func TestCleanupAndRetention(t *testing.T) {
-	database := func(t *testing.T) *icingadb.DB {
-		rdb := getEmptyDatabase(t)
-		db, err := sqlx.Connect(rdb.Driver(), rdb.DSN())
-		require.NoError(t, err, "connecting to database")
-		t.Cleanup(func() { _ = db.Close() })
-
-		return icingadb.NewDb(db, logging.NewLogger(it.Logger(t).Sugar(), time.Second), &icingadb.Options{})
-	}
-	start := time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC)
-	startMilli := start.UnixMilli()
-
-	scenarios := []struct {
-		limit        int64
-		rowsToDelete int64
-		rowsToSpare  int64
-	}{
-		{10, 9, 100},
-		{10, 10, 100},
-		{10, 100, 100},
-		{10, 1000, 100},
-		{100, 100, 100},
-		{1000, 100000, 100},
+	rdb := getDatabase(t)
+	db, err := sqlx.Open(rdb.Driver(), rdb.DSN())
+	require.NoError(t, err, "connecting to SQL database shouldn't fail")
+	t.Cleanup(func() { _ = db.Close() })
+
+	var schema string
+	switch db.DriverName() {
+	case "mysql":
+		schema = `SCHEMA()`
+	case "postgres":
+		schema = `CURRENT_SCHEMA()`
 	}
 
-	for i := range scenarios {
-		scenario := scenarios[i]
-
-		t.Run(fmt.Sprintf("%d rows with limit %d", scenario.rowsToDelete, scenario.limit), func(t *testing.T) {
-			t.Parallel()
-
-			table := fmt.Sprintf("retention_test_limit_%d_delete_%d", scenario.limit, scenario.rowsToDelete)
-
-			db := database(t)
-
-			_, err := db.Exec(fmt.Sprintf(`CREATE TEMPORARY TABLE %s (id int NOT NULL, event_time bigint NOT NULL, PRIMARY KEY (id))`, table))
-			require.NoError(t, err)
+	reten := retention{
+		Days: 7,
+		Options: map[string]int{
+			"acknowledgement": 0, // No retention
+			"comment":         1,
+			"downtime":        2,
+			// notification and state default to 7
+		},
+	}
 
-			_, err = db.Exec(fmt.Sprintf(`CREATE INDEX idx_%[1]s ON %[1]s (event_time)`, table))
+	rowsToDelete := 10000
+	rowsToSpare := 1000
+
+	for category, stmt := range retentionStatements {
+		rows, err := db.Query(db.Rebind(fmt.Sprintf(`
+SELECT column_name
+FROM information_schema.columns
+WHERE table_schema = %s AND table_name = ? AND column_name NOT IN (?, ?) AND is_nullable = ?`,
+			schema)),
+			stmt.Table, stmt.PK, stmt.Column, "NO")
+		if err != nil {
```

Comment:
I'd move everything that's related to fixing up the schema for the test into a helper function. Either it gets a pretty descriptive name or a comment explaining why that's done would also be nice.

(Comment affect more code than selected but I didn't want to select half the diff because that would also be pointless.)

Reply:
Done.